### PR TITLE
drivers: Add support of TCA6507

### PR DESCRIPTION
**Pull request by wysman**
### Contribution description

Add support for the TCA6507 chip in RIOT.

The TCA6507 is a LED driver over I2C which allow to turn ON/OFF, Blinking, and Fading at Programmable Rates. More detailled description can be found in the following datasheet.
http://www.ti.com/lit/ds/symlink/tca6507.pdf

No evaluation kit for TCA6507 is available, you must a have a custom board if you want to test it.

### Testing procedure

This module can be tested with any board with I2C support.

This code have been tested on a custom board based on the ST B-L072Z-LRWAN1 board.

### CODE REVIEW

File: drivers/include/tca6507.h [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+ /**
+  * @ingroup     drivers_actuators
+  * @ingroup     drivers_saul
+  * @defgroup    drivers_tca6507 TCA6507 I2C Led controller
+  * @brief       Internal definitions for TCA6507 I2C Led controller
```
Comment (aabadie) [inline drivers/include/tca6507.h]: This is the public API.

**Comment by wysman** [comment]
Hi @aabadie,

Thanks you for your quick answer about this PR.
I will write a `tests/driver_tca6507` and provide a photo /  GIF.

File: drivers/include/tca6507.h (hunk 1 above)
Comment (wysman) [reply drivers/include/tca6507.h]: Copy / Paste issue !

**Comment by wysman** [comment]
I have write a simple test application for the driver, it's drive 3 leds. The first one is in automatic blink mode with a heartbeat pattern, and two others are drive manually using set/clear. I provide a GIF to show the execution of the test aplication.

https://gph.is/g/EGqJDzn

Output of the application : tests/driver_tca6507
```
2019-10-15 14:32:23,761 - INFO # main(): This is RIOT! (Version: 2019.10-devel-1121-g93338-drivers.tca6507)
2019-10-15 14:32:23,763 - INFO # TCA6507 LED Driver: test application
2019-10-15 14:32:23,763 - INFO # [OK]
2019-10-15 14:32:23,764 - INFO # TCA6507 LED Driver: Clear all leds
2019-10-15 14:32:23,764 - INFO # [OK]
2019-10-15 14:32:23,765 - INFO # TCA6507 LED Driver: Configure Bank 0 at 20%
2019-10-15 14:32:23,765 - INFO # [OK]
2019-10-15 14:32:23,766 - INFO # TCA6507 LED Driver: Configure Bank 1 at 100%
2019-10-15 14:32:23,766 - INFO # [OK]
2019-10-15 14:32:23,766 - INFO # TCA6507 LED Driver: Configure fade on time at 64ms
2019-10-15 14:32:23,766 - INFO # [OK]
2019-10-15 14:32:23,767 - INFO # TCA6507 LED Driver: Configure fade off time at 64ms
2019-10-15 14:32:23,767 - INFO # [OK]
2019-10-15 14:32:23,768 - INFO # TCA6507 LED Driver: Configure fully on time at 128ms
2019-10-15 14:32:23,768 - INFO # [OK]
2019-10-15 14:32:23,769 - INFO # TCA6507 LED Driver: Configure first fully off time at 256ms
2019-10-15 14:32:23,769 - INFO # [OK]
2019-10-15 14:32:23,781 - INFO # TCA6507 LED Driver: Configure second fully off time at 256ms
2019-10-15 14:32:23,781 - INFO # [OK]
2019-10-15 14:32:23,782 - INFO # TCA6507 LED Driver: Blink on led 1 at 20% (i.e. Bank 0)
2019-10-15 14:32:23,782 - INFO # [OK]
2019-10-15 14:32:24,787 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 1
2019-10-15 14:32:24,788 - INFO # [OK]
2019-10-15 14:32:25,043 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 2
2019-10-15 14:32:25,043 - INFO # [OK]
2019-10-15 14:32:26,064 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 2
2019-10-15 14:32:26,065 - INFO # [OK]
2019-10-15 14:32:26,320 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 1
2019-10-15 14:32:26,320 - INFO # [OK]
2019-10-15 14:32:26,575 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 1
2019-10-15 14:32:26,576 - INFO # [OK]
2019-10-15 14:32:26,831 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 2
2019-10-15 14:32:26,831 - INFO # [OK]
2019-10-15 14:32:27,836 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 2
2019-10-15 14:32:27,837 - INFO # [OK]
2019-10-15 14:32:28,092 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 1
2019-10-15 14:32:28,106 - INFO # [OK]
2019-10-15 14:32:28,361 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 1
2019-10-15 14:32:28,362 - INFO # [OK]
2019-10-15 14:32:28,618 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 2
2019-10-15 14:32:28,618 - INFO # [OK]
2019-10-15 14:32:29,624 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 2
2019-10-15 14:32:29,625 - INFO # [OK]
2019-10-15 14:32:29,880 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 1
2019-10-15 14:32:29,880 - INFO # [OK]
2019-10-15 14:32:30,135 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 1
2019-10-15 14:32:30,135 - INFO # [OK]
2019-10-15 14:32:30,390 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 2
2019-10-15 14:32:30,391 - INFO # [OK]
2019-10-15 14:32:31,411 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 2
2019-10-15 14:32:31,412 - INFO # [OK]
2019-10-15 14:32:31,667 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 1
2019-10-15 14:32:31,668 - INFO # [OK]
2019-10-15 14:32:31,923 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 1
2019-10-15 14:32:31,923 - INFO # [OK]
2019-10-15 14:32:32,178 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 2
2019-10-15 14:32:32,179 - INFO # [OK]
2019-10-15 14:32:33,184 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 2
2019-10-15 14:32:33,184 - INFO # [OK]
2019-10-15 14:32:33,439 - INFO # TCA6507 LED Driver: Toggle Bank 1 for led 1
2019-10-15 14:32:33,440 - INFO # [OK]
2019-10-15 14:32:33,694 - INFO # TCA6507 LED Driver: Clear all leds
2019-10-15 14:32:33,695 - INFO # [OK]
2019-10-15 14:32:33,710 - INFO # TCA6507 LED Driver: End
2019-10-15 14:32:33,710 - INFO # 
```

File: drivers/include/tca6507.h [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+int tca6507_init(tca6507_t *dev, const tca6507_params_t *params);
+
+/**
+ * @brief   Turn on a led
+ *
+ * @param[in] dev           device descriptor
+ * @param[in] led           The led index
+ * @param[in] bank          The bank to use (intensity)
+ *
+ * @return                  TCA6507_OK on successful initialization
```
Comment (aabadie) [inline drivers/include/tca6507.h]: All functions in the API returns `TCA6507_OK` *on successful initialization*. Copy-paste ? :D

File: drivers/tca6507/tca6507.c [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    i2c_acquire(TCA6507_I2C);
+
+    if (i2c_read_regs(TCA6507_I2C, TCA6507_ADDR,
+                      TCA6507_SELECT0 | TCA6507_COMMAND_AUTOINC, &regs,
+                      sizeof(regs), 0) != 0) {
+        DEBUG("[ERROR] Cannot read TCA6507 from I2C.\n");
+        goto finish;
+    }
+
+    // only work if led was is state 000 or 010/110 before.
```
Comment (aabadie) [inline drivers/tca6507/tca6507.c]: Please use C-style comments (and verify if there aren't others like this)

File: tests/driver_tca6507/main.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    rc = tca6507_init(&dev, &tca6507_params[0]);
+    CHECK(rc);
+
+    /* Shutdown all leds */
+    puts("TCA6507 LED Driver: Clear all leds");
+    rc = tca6507_clear_all(&dev);
+    CHECK(rc);
+
+    /* Configure Brightness */
+    puts("TCA6507 LED Driver: Configure Bank 0 at 20\%");
```
Comment (aabadie) [inline tests/driver_tca6507/main.c]:
```suggestion
    puts("TCA6507 LED Driver: Configure Bank 0 at 20%%");
```

File: drivers/include/tca6507.h [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+extern "C" {
+#endif
+
+/**
+ * @brief Driver return codes
+ */
+enum {
+    TCA6507_OK,                 /**< All OK */
+    TCA6507_ERR_NODEV,          /**< No valid device found on I2C bus */
+    TCA6507_ERR_I2C             /**< An error occured when reading/writing on I2C bus */
```
Comment (aabadie) [inline drivers/include/tca6507.h]: s/occured/occurred/

Comment (aabadie) [reply drivers/include/tca6507.h]: It was also reported by codespell.

**Comment by wysman** [comment]
@aabadie, I have integrate your review in the code, and squash the code

File: drivers/tca6507/tca6507.c [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+        regs[0] = regs[0] & ~mask;
+        regs[1] = regs[1] | mask;
+        regs[2] = regs[2] | mask;
+    }
+    else {
+        regs[0] = regs[0] | mask;
+        regs[1] = regs[1] | mask;
+        regs[2] = regs[2] | mask;
+    }
+
```
Comment (benpicco) [inline drivers/tca6507/tca6507.c]: A comment on these things would be nice - I was wondering whether you forgot a `& ~mask;` in the second case.

**Comment by benpicco** [comment]
Please add a comment to the documentation on how the `bank` property is supposed to be used.

File: drivers/tca6507/tca6507.c (hunk 6 above)
Comment (wysman) [reply drivers/tca6507/tca6507.c]: The code seems correct.
I will add a datasheet extract to explain the content of register 0-2.

**Comment by benpicco** [comment]
So all LEDs in one bank have the same brightness, but you can assign LEDs to banks individually?
Makes sense I guess.

You can squash.

**Comment by wysman** [comment]
@benpicco It's done

File: drivers/tca6507/tca6507.c [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+                       0) != 0) {
+        DEBUG("[ERROR] Cannot read TCA6507 from I2C.\n");
+        goto finish;
+    }
+
+    if (bank == 0) {
+        value = (value & 0xF0) | (payload & 0x0F);
+    }
+    else {
+        value = (value & 0x0F) | ((payload & 0x0F) << 4);
```
Comment (benpicco) [inline drivers/tca6507/tca6507.c]: And here

**Comment by wysman** [comment]
Can i run Murdock on my debian 10 for local tests ?

**Comment by benpicco** [comment]
> Can i run Murdock on my debian 10 for local tests ?

sure, after all it's just a script that builds all the tests and examples for all boards.
@kaspar030 can probably point you in the right direction.

But tbh, while it's a laudable goal to go easy on our CI server, most people will just push their changes and [wait for the results](https://ci.riot-os.org/).

**Comment by benpicco** [comment]
The change looks good, replacing magic numbers with defines is always a good idea.
Travis found some [issues with the documentation](https://travis-ci.org/RIOT-OS/RIOT/builds/611821913) though.

File: drivers/include/tca6507.h [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+ */
+int tca6507_intensity(const tca6507_t *dev, uint8_t intensity, uint8_t bank);
+
+/**
+ * @brief   Configure the master intensity
+ *
+ * @param[in] dev           device descriptor
+ * @param[in] intensity     Intensity of the led in percent
+ * @param[in] bank          The master intensity, this brightness can be configured
+ *                          to overwrite the bank brightness
```
Comment (benpicco) [inline drivers/include/tca6507.h]: `tca6507_master_intensity()` only takes `brightness` as a parameter.

File: drivers/tca6507/tca6507.c [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+                    TCA6507_SELECT0 | TCA6507_COMMAND_AUTOINC, &regs,
+                    sizeof(regs), 0) != 0) {
+      DEBUG("[ERROR] Cannot read TCA6507 (I2C)\n");
+      goto finish;
+  }
+  ret = TCA6507_OK;
+
+  puts("tca6507 registers :");
+  for (size_t i=0; i<sizeof(regs); i++) {
+    printf("\t%02x: %02x\n", i, regs[i]);
```
Comment (benpicco) [inline drivers/tca6507/tca6507.c]:
```suggestion
    printf("\t%02lx: %02x\n", i, regs[i]);
```

File: drivers/include/tca6507.h (hunk 8 above)
Comment (wysman) [reply drivers/include/tca6507.h]: Fixed

File: drivers/tca6507/tca6507.c [hunk 10]
```diff
@@ -377,7 +377,7 @@ int tca6507_dump_registers(const tca6507_t *dev)
 
   puts("tca6507 registers :");
   for (size_t i=0; i<sizeof(regs); i++) {
-    printf("\t%02x: %02x\n", i, regs[i]);
+    printf("\t%02lx: %02x\n", i, regs[i]);
```
Comment (benpicco) [inline drivers/tca6507/tca6507.c]:
```suggestion
    printf("\t%02x: %02x\n", (unsigned)i, regs[i]);
```

Sorry, should have read the [coding guidelines](https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#-wformat) on that.

**Comment by miri64** [comment]
Ping @wysman?

**Comment by wysman** [comment]
@miri64 I don't have the board anymore to make extra tests, so i can not works on it !
Last murdock remarks are trivial, i let's you handle that before merge.

**Pull request closed without merging**